packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -303,26 +286,124 @@ export class OdspDocumentService implements IDocumentService {
         });
     }
 
+    private clearJoinSessionTimer() {
+        if (this.joinSessionRefreshTimer !== undefined) {
+            clearTimeout(this.joinSessionRefreshTimer);
+            this.joinSessionRefreshTimer = undefined;
+        }
+    }
+
+    private async scheduleJoinSessionRefresh(delta: number) {
```

Comment:
I'm surprised to see that compiler does not complain that caller of this API does not need to handle exceptions (do .catch() on returned promise). I'd suggest making this function fire & forget (void, not async) and either handle rejection here or add "void" in front of promise to tell compiler we do not care (as there is no way for this promise to be rejected)

Reply:
I am catching the error when making the call to scheduleJoinSessionRefresh. I am not seeing any issue in keeping this async and then catching any error and reporting to telemetry.

Reply:
But setTimeout takes void callback. And you have async callback, which basically means that any exceptions thrown become global unhandled promises.

Reply:
Basically, if you want to propagate exceptions to callers of scheduleJoinSessionRefresh, then code should look like this:

```ts
 private async scheduleJoiawait new Promise<void>((resolve, reject) => {
            this.joinSessionRefreshTimer = setTimeout(() => {
                getWithRetryForTokenRefresh(async (options) => {nSessionRefresh(delta: number) {
 await this.joinSession(false, options);
                    resolve();
                }).catch(error) { reject(error); };
            }, delta);
        });
    }
```

Reply:
sorry, this is more about next comment

Reply:
Got it. I have a question, since we are not awaiting getWithRetryForTokenRefresh, how can we be sure that this call will be executed immediately after the timer goes off?

Reply:
I'm not 100% sure what you ask, but I'll try to answer.
In general, OS itself (and by virtue - JS environment) do not provide you here any strong guarantees. Timers can be delayed, coalesced (many timers within some range fire at the same time, sooner or later than what you specified), plus system can be so busy that it can't schedule thread to run (timer storm).
But in most cases callback will run pretty quickly, as priority of a thread that get awaited is high. JS engine itself will do similar kind of balancing and if JS there is stack unrolled (i.e. there is no code that burns code in infinite loop), then timer callback will eventually run.

Continuations in JS are micro-tasks. I.e. if task itself (here - timer callback) run and completed, then continuations will run right away, before running any tasks.
If one of the continuations awaits on some IO (network), then co-routine (i.e. full context with stack and local state) is saved and switched to run some other co-routine, but once network call completes, it gets scheduled and runs the moment there is no task on the stack (no infinite loop task).

Hope that helps

Reply:
Thanks.